cirq/study/resolver.py

```diff
@@ -244,4 +244,9 @@ def _sympy_pass_through(val: Any) -> Optional[Any]:
         return val.p / val.q
     if val == sympy.pi:
         return np.pi
+
+    getter = getattr(val, '_sympy_pass_through_', None)
```

Comment:
I think it'd be helpful to change the name, because the point of defining a special method is to broaden this beyond the hard-coded sympy handling. In addition, the code above allows transforming values, not just passing them through (e.g. sympy integers become plain old integers). So maybe we could call the special method something like `_resolver_value_` and if it exists we call it to get a value to pass through, which could be different from `val` itself. The method could return `NotImplemented` to indicate that the value should _not_ be passed through. We could also change this function itself to use `NotImplemented` as the sentinel value instead of `None` if we want to allow things to resolve to `None` (the function is only called in `value_of` above, so that change wouldn't affect any other code).

Reply:
> We could also change this function itself to use NotImplemented as the sentinel value instead of None if we want to allow things to resolve to None (the function is only called in value_of above, so that change wouldn't affect any other code).

I tried this but it turns out that we're relying on `None` as a sentinel in some subtle way that I couldn't quite unravel, probably because of this line: https://github.com/quantumlib/Cirq/blob/master/cirq/study/resolver.py#L175

Reply:
That `None` was me :p A valid workaround would be to replace `None` with a file-local class used only for flagging potential loops in recursive parameter evaluation.

Reply:
Hmm...I guess replacing wouldn't resolve this entirely, since `None` is checked for earlier in `value_of`.

Reply:
@95-martin-orion thanks for the hint, got all the tests to pass now, check if this is what you had in mind?

Reply:
I'd put a leading underscore on `RecursionFlag` just to make it abundantly clear that it's a non-public object, but otherwise this looks right.

Reply:
+1 to leading underscore. I'd also suggest adding a comment that this is used to mark values that are being resolved recursively to detect loops.